supabase-openai-summary/src/steps/summarizers.py

```diff
@@ -0,0 +1,75 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+from typing import List
+import openai
+
+from zenml.steps import step, BaseParameters, StepContext
+from zenml.client import Client
+from zenml.post_execution import get_pipeline, PipelineRunView
+
+
+class SummarizerParams(BaseParameters):
+    """Prompts for the summarizer."""
+
+    system_content: str = "Act like a data analytics expert."
+    prompt_preamble: str = "Summarize the latest data from the database. The data is YouTube video titles and we are keen to understand what sort of videos users are accessing, the trends, and similarities between them. We also have the analysis from last time that we will give to you as input."
+    prompt_example: str = "The latest data indicates the following key insights: 1. The number of users has increased by 10% in the last 24 hours. 2. The number of users has increased by 10% in the last 24 hours. 3... Compared to our last summary, it seems our users are starting to watch more videos about cats."
+
+
+@step(enable_cache=True)
+def gpt_4_summarizer(
+    params: SummarizerParams, documents: List[str], context: StepContext
+) -> str:
+    """Summarizes the data using GPT-4."""
+    openai_secret = Client().get_secret("openai")
+
+    p = get_pipeline(context.pipeline_name)
+
+    last_step = PipelineRunView(
+        Client().list_runs(size=5, pipeline_id=p.id)[1]
+    ).get_step("generate_summary")
+    if len(last_step.outputs) == 0:
+        last_analysis = "No previous analysis found."
+    else:
+        last_analysis = last_step.output.read()
```

Comment:
This might not always work as expected since `get_pipeline(NAME)` currently returns the latest version of a pipeline with that name but you might have newer runs under an older version. In https://github.com/zenml-io/zenml/pull/1510 I tried to fix this by making `get_pipeline()` return a new `PipelineClassView` whose `.runs` property contains the runs of all pipeline versions with that name. The PR is not merged yet, but if we wanted to be compatible with that change, you should do it like this instead:

```suggestion
    last_analysis = "No previous analysis found."
    p = get_pipeline(context.pipeline_name)
    if p.runs:
        last_step = p.runs[0].get_step("generate_summary")
        if last_step and len(last_step.outputs) == 1:
            last_analysis = last_step.output.read()
```

Reply:
The version I suggested should already work under the current release and will be backwards compatible, could you try it? If it works, I suggest we change it now, otherwise the project will immediately be broken with the next release

Reply:
@fa9r get_pipelines() doesnt return what I expect. I started with this way but its broken and I hope its fixed in your release. I will upgrade once the zenml is released and pin to 0.38.0 for now